docs/proposals/2170-kubeflow-training-v2/README.md

```diff
@@ -279,6 +279,43 @@ type TrainJob struct {
 	Status TrainJobStatus `json:"status,omitempty"`
 }
 
+const (
+	// TrainJobSuspended means the TrainJob is suspended.
+	TrainJobSuspended string = "Suspended"
+
+	// TrainJobCompleted means that the actual jobs have completed its execution.
+	TrainJobCompleted string = "Completed"
+
+	// TrainJobFailed means that the actual jobs have failed its execution.
+	TrainJobFailed string = "Failed"
+
+	// TrainJobCreated means that the actual jobs creation has succeeded.
+	TrainJobCreated string = "Created"
+)
+
+const (
+	// TrainJobSuspendedReason is the "Suspended" condition reason.
+	// When the TrainJob is suspended, this is added.
+	TrainJobSuspendedReason string = "Suspended"
+
+	// TrainJobResumedReason is the "Suspended" condition reason.
+	// When the TrainJob suspension is changed from True to False, this is added.
+	TrainJobResumedReason string = "Resumed"
+
+	// TrainJobJobsCreationSucceededReason is the "Created" condition reason.
+	// When the creating objects succeeded after building succeeded, this is added.
+	TrainJobJobsCreationSucceededReason string = "JobsCreationSucceeded"
+
+	// TrainJobJobsBuildFailedReason is the "Created" condition reason.
+	// When the building objects based on the TrainJob and the specified runtime failed,
+	// this is added.
+	TrainJobJobsBuildFailedReason string = "JobsBuildFailed"
+
+	// TrainJobJobsCreationFailedReason is "Created" condition reason.
+	// When the creating objects failed even though building succeeded, this is added.
+	TrainJobJobsCreationFailedReason string = "JobsCreationFailed"
```

Comment:
Should we introduce those conditions in the 2nd iteration ?
I think, we should discuss if users really want to decouple conditions when TrainJob's object creation failed and when TrainJob fails.

Reply:
These are reasons, so we use these in the following. Additionally, if we do not distinguish `JobCreationFailed` and `JobBuildFailed`, it is hard to understand which part failed. So these reasons must be contained in the first phase.

- JobsBuildFailed: https://github.com/kubeflow/training-operator/blob/9e04bdd74920cbe12ecc786010d80fb8986f27bd/pkg/controller.v2/trainjob_controller.go#L71
- JobsCreationFailed: https://github.com/kubeflow/training-operator/blob/9e04bdd74920cbe12ecc786010d80fb8986f27bd/pkg/controller.v2/trainjob_controller.go#L98

```yaml
status:
  conditions:
  - type: Created
    status: false
    reason: JobsCreationFailed
```

```yaml
status:
  conditions:
  - type: Created
    status: false
    reason: JobsBuildFailed
```

```yaml
status:
  conditions:
  - type: Created
    status: true
    reason: JobsCreationSucceeded
```

Reply:
Since we have validation webhook, what is the use-case you see when we can hit the JobsBuildFailed reason ?

Reply:
So, I guess one of the examples could be if NewObject calls Kubernetes API server and this call fails.
In that case, we want to transition TrainJob to JobsBuildFailed status.

Reply:
> In that case, we want to transition TrainJob to JobsBuildFailed status.

Yes, that's right. In the current JobSet plugin, the building error could happen in the following:

- https://github.com/kubeflow/training-operator/blob/fd4d102f95af8c918f76d9d149dc3a6557d0512f/pkg/runtime.v2/framework/plugins/jobset/jobset.go#L93
- https://github.com/kubeflow/training-operator/blob/fd4d102f95af8c918f76d9d149dc3a6557d0512f/pkg/runtime.v2/framework/plugins/jobset/jobset.go#L98

Additionally, every plugin could return errors during Kubeflow Job Pipeline Framerowk.